Mathlib/LinearAlgebra/InvariantBasisNumber.lean

```diff
@@ -302,20 +304,28 @@ private def induced_equiv [Fintype ι'] (I : Ideal R) (e : (ι → R) ≃ₗ[R]
 
 end
 
-section
-
-attribute [local instance] Ideal.Quotient.field
-
-/-- Nontrivial commutative rings have the invariant basis number property.
+/-- Nontrivial commutative semirings satisfy the rank condition.
 
 In fact, any nontrivial commutative ring satisfies the strong rank condition, see
-`commRing_strongRankCondition`. We prove this instance separately to avoid dependency on
-`LinearAlgebra.Charpoly.Basic`. -/
-instance (priority := 100) invariantBasisNumber_of_nontrivial_of_commRing {R : Type u} [CommRing R]
-    [Nontrivial R] : InvariantBasisNumber R :=
-  ⟨fun e =>
-    let ⟨I, _hI⟩ := Ideal.exists_maximal R
-    eq_of_fin_equiv (R ⧸ I)
-      ((Ideal.piQuotEquiv _ _).symm ≪≫ₗ induced_equiv _ e ≪≫ₗ Ideal.piQuotEquiv _ _)⟩
-
-end
+`commRing_strongRankCondition`, but it is not clear whether this generalizes to semirings.
+We prove this instance separately to avoid dependency on `LinearAlgebra.Charpoly.Basic`. -/
+instance (priority := 100) rankCondition_of_nontrivial_of_commSemiring {R : Type*}
+    [CommSemiring R] [Nontrivial R] : RankCondition R where
+  le_of_fin_surjective {n m} f hf := by
+    by_contra! lt
+    let p : (Fin m → R) →ₗ[R] Fin n → R :=
+      ⟨⟨(· ∘ Fin.castLE lt.le), fun _ _ ↦ rfl⟩, fun _ _ ↦ rfl⟩
+    have hp : Surjective p := fun x ↦ ⟨(Fin.castLE lt.le).extend x 0,
+      funext fun _ ↦ (Fin.castLE_injective lt.le).extend_apply ..⟩
+    have ⟨g, eq⟩ := (p ∘ₗ f).exists_rightInverse_of_surjective
+      (LinearMap.range_eq_top.mpr <| hp.comp hf)
+    apply_fun Matrix.toLinAlgEquiv'.symm at eq
+    rw [← LinearMap.mul_eq_comp, ← LinearMap.one_eq_id, map_mul, map_one,
+      Matrix.mul_eq_one_comm_of_equiv (Equiv.refl _), ← map_mul,
+      ← map_one Matrix.toLinAlgEquiv'.symm, Matrix.toLinAlgEquiv'.symm.injective.eq_iff] at eq
+    have := congr_fun ((p.coe_comp f ▸ LinearMap.injective_of_comp_eq_id _ _ eq).of_comp_right hf
```

Comment:
Can you annotate what type this `this` has?

Reply:
Instead annotating the whole thing, I think it's better to annotate the part that says `p` is injective.

Reply:
Let me also extract the lemma that is the converse to [Function.Surjective.injective_comp_right](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Function/Basic.html#Function.Surjective.injective_comp_right) ...

Reply:
Extracted the lemma, which required importing Nontrivial into Function.Basic. Also generalized Nontrivial to Sort* (Subsingleton already worked for Sort*).